// india-states-districts/index.js
const fs = require('fs');
const path = require('path');

// Load the JSON data
const dataPath = path.join(__dirname, 'state_discripts.json');
const rawData = fs.readFileSync(dataPath, 'utf8');
const data = JSON.parse(rawData);

/**
 * Get all states from India (including union territories)
 * @returns {Array<string>} Array of state names and union territories
 */
function getAllStates() {
  const states = data.states.map(state => state.name);
  const unionTerritories = data.union_territories.map(ut => ut.name);
  return [...states, ...unionTerritories];
}

/**
 * Get districts for a given state or union territory
 * @param {string} stateName - The name of the state or union territory
 * @returns {Array<string>|null} Array of district names or null if state/union territory not found
 */
function getDistrictsByState(stateName) {
  if (!stateName || typeof stateName !== 'string') {
    return null;
  }

  // Search in states
  const state = data.states.find(
    s => s.name.toLowerCase() === stateName.toLowerCase()
  );
  
  if (state) {
    return state.districts;
  }

  // Search in union territories
  const unionTerritory = data.union_territories.find(
    ut => ut.name.toLowerCase() === stateName.toLowerCase()
  );

  if (unionTerritory) {
    return unionTerritory.districts;
  }

  return null;
}

/**
 * Get all states with their districts (including union territories)
 * @returns {Array<{name: string, districts: Array<string>}>} Array of objects containing state/union territory name and districts
 */
function getAllStatesWithDistricts() {
  const states = data.states.map(state => ({
    name: state.name,
    districts: state.districts
  }));
  const unionTerritories = data.union_territories.map(ut => ({
    name: ut.name,
    districts: ut.districts
  }));
  return [...states, ...unionTerritories];
}

/**
 * Check if a state or union territory exists
 * @param {string} stateName - The name of the state or union territory
 * @returns {boolean} True if state/union territory exists, false otherwise
 */
function stateExists(stateName) {
  if (!stateName || typeof stateName !== 'string') {
    return false;
  }

  const allStates = getAllStates();
  return allStates.some(
    state => state.toLowerCase() === stateName.toLowerCase()
  );
}

module.exports = {
  getAllStates,
  getDistrictsByState,
  getAllStatesWithDistricts,
  stateExists
};
